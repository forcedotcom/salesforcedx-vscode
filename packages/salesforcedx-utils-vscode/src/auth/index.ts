/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

/**
 * This file is intentionally left blank. AuthUtil is exported
 * in the root of salesforcedx-utils-vscode.
 */
