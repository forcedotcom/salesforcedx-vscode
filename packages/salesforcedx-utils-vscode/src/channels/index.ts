/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

import { ChannelService } from './channelService';
export const channelService = ChannelService.getInstance();
export { DEFAULT_SFDX_CHANNEL } from './channelService';
export { ChannelService } from './channelService';
