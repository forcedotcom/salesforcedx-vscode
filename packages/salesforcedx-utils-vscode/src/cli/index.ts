/*
 * Copyright (c) 2017, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export interface Command {
  readonly command: string;
  readonly description?: string;
  readonly args: string[];
  readonly logName?: string;

  toString(): string;
  toCommand(): string;
}

export { CommandBuilder, SfdxCommandBuilder } from './commandBuilder';
export {
  CliCommandExecutor,
  CliCommandExecution,
  CommandExecution,
  CompositeCliCommandExecutor,
  GlobalCliEnvironment
} from './commandExecutor';
export { CommandOutput } from './commandOutput';
export { ForceConfigGet } from './forceConfigGet';
export { ForceOrgDisplay, OrgInfo } from './forceOrgDisplay';
export { LocalCommandExecution } from './localCommandExecutor';
export {
  ForceDeployResultParser,
  ForceSourceDeployErrorResponse,
  ForceSourceDeploySuccessResponse,
  DeployResult,
  CONFLICT_ERROR_NAME
} from './deployResultParser';
export { TestRunner } from './testRunner';
export {
  OrgCreateSuccessResult,
  OrgCreateErrorResult,
  OrgCreateResultParser
} from './orgCreateResultParser';
export {
  OrgOpenSuccessResult,
  OrgOpenContainerResultParser,
  OrgOpenErrorResult
} from './orgOpenContainerResultParser';
export {
  DiffErrorResponse,
  DiffResultParser,
  DiffSuccessResponse
} from './diffResultParser';
