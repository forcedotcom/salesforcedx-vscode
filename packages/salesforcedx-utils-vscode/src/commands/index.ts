/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export { ChannelService } from './channelService';

import { NotificationService } from './notificationService';
export const notificationService = NotificationService.getInstance();
export { NotificationService };
export { ProgressNotification } from './progressNotification';
