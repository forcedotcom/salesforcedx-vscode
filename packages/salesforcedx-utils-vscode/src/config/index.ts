/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

/**
 * This file is intentionally left blank. ConfigSource and ConfigUtil are exported
 * in the root of salesforcedx-utils-vscode.
 */
