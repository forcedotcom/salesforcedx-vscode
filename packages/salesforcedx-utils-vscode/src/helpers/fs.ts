/*
 * Copyright (c) 2025, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import { dirname } from 'node:path';
import * as vscode from 'vscode';

export const readFile = async (filePath: string): Promise<string> => {
  try {
    const uri = vscode.Uri.file(filePath);
    const data = await vscode.workspace.fs.readFile(uri);
    return Buffer.from(data).toString('utf8');
  } catch (error) {
    throw new Error(`Failed to read file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Writes content to a file
 * @param filePath The path to the file
 * @param content The content to write
 */
export const writeFile = async (filePath: string, content: string): Promise<void> => {
  try {
    const dirPath = dirname(filePath);
    if (!(await fileOrFolderExists(dirPath))) {
      await createDirectory(dirPath);
    }

    const encoder = new TextEncoder();
    const uint8Array = encoder.encode(content);
    await vscode.workspace.fs.writeFile(vscode.Uri.file(filePath), uint8Array);
  } catch (error) {
    throw new Error(`Failed to write file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Checks if a file exists
 * @param filePath The path to the file
 * @returns True if the file exists, false otherwise
 */
export const fileOrFolderExists = async (filePath: string): Promise<boolean> => {
  try {
    const uri = vscode.Uri.file(filePath);
    await vscode.workspace.fs.stat(uri);
    return true;
  } catch {
    return false;
  }
};

/**
 * Creates a directory recursively.  Will not throw if the directory already exists.
 * @param dirPath The path to the directory
 */
export const createDirectory = async (dirPath: string): Promise<void> => {
  try {
    const uri = vscode.Uri.file(dirPath);
    await vscode.workspace.fs.createDirectory(uri);
  } catch (error) {
    throw new Error(`Failed to create directory ${dirPath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Deletes a file
 * @param filePath The path to the file
 */
export const deleteFile = async (
  filePath: string,
  options: { recursive?: boolean; useTrash?: boolean } = {}
): Promise<void> => {
  try {
    const uri = vscode.Uri.file(filePath);
    await vscode.workspace.fs.delete(uri, options);
  } catch (error) {
    throw new Error(`Failed to delete file ${filePath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Reads a directory's contents
 * @param dirPath The path to the directory
 * @returns Array of file/directory names
 */
export const readDirectory = async (dirPath: string): Promise<string[]> => {
  try {
    const uri = vscode.Uri.file(dirPath);
    const entries = await vscode.workspace.fs.readDirectory(uri);
    return entries.map(([name]) => name);
  } catch (error) {
    throw new Error(`Failed to read directory ${dirPath}: ${error instanceof Error ? error.message : String(error)}`);
  }
};

/**
 * Gets file stats
 * @param filePath The path to the file
 * @returns File stats including size, creation time, and modification time
 */
export const stat = async (filePath: string): Promise<vscode.FileStat> => {
  try {
    const uri = vscode.Uri.file(filePath);
    return await vscode.workspace.fs.stat(uri);
  } catch (error) {
    throw new Error(
      `Failed to get file stats for ${filePath}: ${error instanceof Error ? error.message : String(error)}`
    );
  }
};

export const safeDelete = async (
  filePath: string,
  options?: { recursive?: boolean; useTrash?: boolean }
): Promise<void> => {
  try {
    const uri = vscode.Uri.file(filePath);
    await vscode.workspace.fs.stat(uri);
    await vscode.workspace.fs.delete(uri, options);
  } catch {
    // File doesn't exist or can't be accessed, do nothing
  }
};

/**
 * Ensures the current working directory is set to the project path
 * @param rootWorkspacePath The path to the root workspace
 */
export const ensureCurrentWorkingDirIsProjectPath = async (rootWorkspacePath: string): Promise<void> => {
  if (rootWorkspacePath && process.cwd() !== rootWorkspacePath) {
    try {
      const uri = vscode.Uri.file(rootWorkspacePath);
      await vscode.workspace.fs.stat(uri);
      process.chdir(rootWorkspacePath);
    } catch {
      // Path doesn't exist, do nothing
    }
  }
};
