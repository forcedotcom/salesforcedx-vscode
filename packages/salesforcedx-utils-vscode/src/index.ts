/*
 * Copyright (c) 2020, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */

export * from './commands';
export {
  CompositeParametersGatherer,
  EmptyParametersGatherer
} from './commands/parameterGatherers';
export { EmptyPostChecker } from './commands/postconditionCheckers';
export {
  EmptyPreChecker,
  SfdxWorkspaceChecker
} from './commands/preconditionCheckers';
export {
  SfdxCommandletExecutor,
  LibraryCommandletExecutor
} from './commands/commandletExecutors';
export { SfdxCommandlet } from './commands/sfdxCommandlet';
export { ConfigSource, ConfigUtil } from './config/configUtil';
export {
  SETTING_CLEAR_OUTPUT_TAB,
  SFDX_CORE_CONFIGURATION_NAME,
  SFDX_CORE_EXTENSION_NAME
} from './constants';
export {
  OrgUserInfo,
  WorkspaceContextUtil
} from './context/workspaceContextUtil';
export {
  TelemetryService,
  TelemetryBuilder,
  TelemetryData,
  Properties,
  Measurements
} from './telemetry/telemetry';
export * from './helpers';
export { TraceFlags } from './helpers/traceFlags';
export { TelemetryReporter } from './telemetry/telemetryReporter';
export {
  hasRootWorkspace,
  getRootWorkspace,
  getRootWorkspacePath,
  workspaceUtils
} from './workspaces';
export * from './cli';
export * from './cli/commandExecutor';
export * from './i18n';
export * from './types';
export * from './date';
export * from './output';
export * from './predicates';
export * from './providers';
export * from './services';
