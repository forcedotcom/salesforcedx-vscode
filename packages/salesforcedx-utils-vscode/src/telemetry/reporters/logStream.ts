/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
'use strict';

import { TelemetryReporter } from '@salesforce/vscode-service-provider';
import * as fs from 'fs';
import * as path from 'path';
import { Disposable, workspace } from 'vscode';
import { WorkspaceContextUtil } from '../../context/workspaceContextUtil';

/**
 * Represents a telemetry reporter that logs telemetry events to a file.
 */
export class LogStream extends Disposable implements TelemetryReporter {
  private toDispose: Disposable[] = [];

  private stream: fs.WriteStream | undefined;

  constructor(
    private extensionId: string,
    private logFilePath: string
  ) {
    super(() => this.toDispose.forEach(d => d && d.dispose()));
    logFilePath = path.join(logFilePath, `${this.extensionId}.txt`);
    this.stream = fs.createWriteStream(logFilePath, {
      flags: 'a',
      encoding: 'utf8',
      autoClose: true
    });
    this.toDispose.push(workspace.onDidChangeConfiguration(() => () => {}));
    console.log(
      'VS Code telemetry event logging enabled for: ' +
        this.extensionId +
        '. Telemetry events will be written via write stream to a file at: ' +
        this.logFilePath +
        '.'
    );
  }

  public sendTelemetryEvent(
    eventName: string,
    properties?: { [key: string]: string },
    measurements?: { [key: string]: number }
  ): void {
    const orgId = WorkspaceContextUtil.getInstance().orgId;
    if (orgId && properties) {
      properties.orgId = orgId;
    } else if (orgId) {
      properties = { orgId };
    }

    if (this.stream) {
      this.stream.write(
        `telemetry/${eventName} ${JSON.stringify({
          properties,
          measurements
        })}\n`
      );
    }
  }

  public sendExceptionEvent(
    exceptionName: string,
    exceptionMessage: string,
    measurements?: { [key: string]: number }
  ): void {
    const orgId = WorkspaceContextUtil.getInstance().orgId || '';
    const properties = { orgId };
    console.log(
      'LogStream.sendExceptionEvent - exceptionMessage: ' + exceptionMessage
    );

    if (this.stream) {
      this.stream.write(
        `telemetry/${exceptionName} ${JSON.stringify({
          properties,
          measurements
        })}\n`
      );
    }
  }

  public dispose(): Promise<any> {
    const flushEventsToLogger = new Promise<any>(resolve => {
      if (!this.stream) {
        return resolve(void 0);
      }
      this.stream.on('finish', resolve);
      this.stream.end();
    });

    return flushEventsToLogger;
  }
}
