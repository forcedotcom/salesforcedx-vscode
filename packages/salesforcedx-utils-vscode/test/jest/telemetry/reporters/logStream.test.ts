/*
 * Copyright (c) 2024, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
import * as fs from 'fs';
import * as path from 'path';
import { WorkspaceContextUtil } from '../../../../src';
import { LogStream } from '../../../../src/telemetry/reporters/logStream';

jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  createWriteStream: jest.fn()
}));
const fsMocked = jest.mocked(fs);

describe('LogStream', () => {
  const fakeExtensionId = 'myExtension';
  const fakeLogFilePath = '/path/to/logs';
  const mockWorkspaceContextUtil = {
    onOrgChange: jest.fn(),
    getConnection: jest.fn()
  };
  const mockWrite = jest.fn();
  const mockEnd = jest.fn();
  let workspaceContextUtilGetInstanceSpy: jest.SpyInstance;
  let logStream: LogStream;

  beforeEach(() => {
    workspaceContextUtilGetInstanceSpy = jest
      .spyOn(WorkspaceContextUtil, 'getInstance')
      .mockReturnValue(mockWorkspaceContextUtil as any);

    fsMocked.createWriteStream.mockReturnValue({
      write: mockWrite,
      end: mockEnd
    } as any);
  });

  it('should write the telemetry event to the file', () => {
    const eventName = 'myEvent';
    const properties = { key: 'value' };
    const measurements = { count: 1 };
    logStream = new LogStream(fakeExtensionId, fakeLogFilePath);

    logStream.sendTelemetryEvent(eventName, properties, measurements);

    expect(fsMocked.createWriteStream).toHaveBeenCalledWith(
      `${path.join(fakeLogFilePath, fakeExtensionId)}.txt`,
      expect.any(Object)
    );
    expect((logStream as any).stream).toBeDefined();
    expect(workspaceContextUtilGetInstanceSpy).toHaveBeenCalled();
    expect(mockWrite.mock.calls[0][0]).toMatchSnapshot();
  });

  it('should write the exception event to the file', () => {
    const exceptionName = 'myException';
    const exceptionMessage = 'An exception occurred';
    const measurements = { count: 1 };

    logStream.sendExceptionEvent(exceptionName, exceptionMessage, measurements);

    expect(workspaceContextUtilGetInstanceSpy).toHaveBeenCalled();
    expect(mockWrite.mock.calls[0][0]).toMatchSnapshot();
  });
});
