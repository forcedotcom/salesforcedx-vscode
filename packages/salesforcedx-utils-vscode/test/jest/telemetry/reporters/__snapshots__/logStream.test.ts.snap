// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LogStream should write the exception event to the file 1`] = `
"telemetry/myException {"properties":{"orgId":""},"measurements":{"count":1}}
"
`;

exports[`LogStream should write the telemetry event to the file 1`] = `
"telemetry/myEvent {"properties":{"key":"value"},"measurements":{"count":1}}
"
`;
